File: android/ValleTPV/ValleTPVlib/src/main/java/com/valleapp/valletpvlib/db/entities/Tecla.kt
package com.valleapp.valletpvlib.db.entities

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.valleapp.valletpvlib.interfaces.IBaseEntity
import org.json.JSONObject

@Entity(tableName = "teclas")
data class Tecla(
    @PrimaryKey val ID: Int,
    val Nombre: String,
    val P1: Double,
    val P2: Double,
    val Precio: Double,
    val RGB: String,
    val IDSeccion: Int,
    val Tag: String?,
    val Orden: Int,
    val IDSec2: Int?,
    val tipo: String?,
    val descripcion_t: String?,
    val descripcion_r: String?
) : IBaseEntity<Tecla> {

    override fun entityFromJson(obj: JSONObject): Tecla {
        return Tecla(
            ID = obj.getInt("ID"),
            Nombre = obj.getString("Nombre"),
            P1 = obj.getDouble("P1"),
            P2 = obj.getDouble("P2"),
            Precio = obj.getDouble("Precio"),
            RGB = obj.getString("RGB"),
            IDSeccion = obj.getInt("IDSeccion"),
            Tag = obj.optString("Tag"),
            Orden = obj.getInt("Orden"),
            IDSec2 = obj.optInt("IDSec2"),
            tipo = obj.optString("tipo"),
            descripcion_t = obj.optString("descripcion_t"),
            descripcion_r = obj.optString("descripcion_r")
        )
    }

    override fun jsonFromEntity(entity: Tecla): JSONObject {
        val json = JSONObject()
        json.put("ID", entity.ID)
        json.put("Nombre", entity.Nombre)
        json.put("P1", entity.P1)
        json.put("P2", entity.P2)
        json.put("Precio", entity.Precio)
        json.put("RGB", entity.RGB)
        json.put("IDSeccion", entity.IDSeccion)
        json.put("Tag", entity.Tag)
        json.put("Orden", entity.Orden)
        json.put("IDSec2", entity.IDSec2)
        json.put("tipo", entity.tipo)
        json.put("descripcion_t", entity.descripcion_t)
        json.put("descripcion_r", entity.descripcion_r)
        return json
    }
}

File: javascript/valle-tpv-root/valle-tpv-lib/src/models/tecla.ts
export interface TeclaParams {
  ID: number;
  nombre: string;
  IDFamilia?: number | null;
  IDSeccion?: number | null;
  IDSeccionCom?: number | null;
  Precio?: number | null ;
  RGB?: string | null;
  p1: number;
  p2: number;
  tag?: string | null;
  orden?: number | null;
  tipo?: string | null;
  descripcion_t: string;
  descripcion_r: string;
  IDParentTecla?:  number | null;
  hay_existencias: 0 | 1;
}

class Tecla {
  ID: number;
  nombre: string;
  IDFamilia?: number | null = null;
  IDSeccion?: number | null = null;
  IDSeccionCom?: number | null = null;
  Precio?: number | null = null;
  RGB?: string | null = null;
  p1: number;
  p2: number;
  tag?: string | null = null;
  orden?: number | null = null;
  tipo?: string | null = null;
  descripcion_t: string;
  descripcion_r: string;
  IDParentTecla?:  number | null = null;
  hay_existencias: 0 | 1 = 1;
  static schema: string | null = '++ID, nombre, p1, p2, IDFamilia, IDParentTecla, IDSeccion, IDSeccionCom, Precio, RGB, tag, orden, tipo, descripcion_t, descripcion_r, hay_existencias';

  constructor({ ID, nombre, p1, p2, descripcion_t, descripcion_r,  tag = null, orden = null, tipo = null,   hay_existencias = 1, IDFamilia = 0, IDParentTecla = null, IDSeccion = 0, IDSeccionCom = 0, Precio = 0, RGB = '' }: TeclaParams) {
    this.ID = ID || Date.now();
    this.nombre = nombre;
    this.p1 = p1;
    this.p2 = p2;
    this.IDFamilia = IDFamilia;
    this.IDSeccion = IDSeccion;
    this.IDSeccionCom = IDSeccionCom;
    this.Precio = Precio;
    this.RGB = RGB;
    this.tag = tag;
    this.orden = orden;
    this.tipo = tipo;
    this.descripcion_t = descripcion_t;
    this.descripcion_r = descripcion_r;
    this.IDParentTecla = IDParentTecla;
    this.hay_existencias = hay_existencias;
  }

}

export default Tecla;
